### PR TITLE
feat(developer): add `&displayMap` system store 🗜

### LINKED ISSUE #1070: Solve problem with rendering diacritics on OSK key caps
Many fonts/platforms fail to render an unattached diacritic on the on-screen keyboard. The existing solutions are:
1) Use a bitmap. This scales poorly and is not supported on some platforms.
2) Create a custom font just for this purpose. This works well, but is a ton of work.
We need a good cross-platform solution that has a low enough bar that a reasonable developer can achieve it.

### LINKED ISSUE #9031: feat(common): support custom OSK font to avoid unattached mark rendering inconsistencies across platforms
**Note:** implementation discussed here diverges from the design we arrived at; for that, see the implementing PR #9032.

## Problem Description

The Keyman On Screen Keyboard frequently has problems displaying diacritic marks and other characters that renderers expect to be combined with a base. The rendering tends to be inconsistent across platforms with the following issues encountered:
* Diacritic marks displaying off the key due to renderer expecting them to be placed over previous letter
* Inconsistent display of dotted circle indicating an unattached mark - some renderers automatically insert a dotted circle for unattached marks, others don't
* Multiple dotted circles displaying on some platforms (especially if using 25CC to try and mitigate the issue)
* Multiple combining marks not rendering correctly together without a base

## Affected Scripts

These issues have been observed for many scripts – even Latin script, for example when combining multiple diacritics onto a single key.

We have documented issues with the following scripts, so far:
* Lao
* Khmer
* Thai
* Hebrew
* Latin
* Tai Tham (see example below)

image

However, this is a general issue for any script which uses combining marks of any kind.

## Attempted Mitigations

* Keyboard authors often attempt to use 25CC as a base. However, renderers are still inconsistent with how they render this.
* Some keyboard authors have used bitmap images – but these render poorly when scaled and are difficult to maintain.
* Talking to renderer maintainers – long-term, this may be fruitful but is unlikely to result in 100% consistent display across all platforms.

### Summary

Create a font for each affected script which has additional glyphs in the PUA, and use these marks in place of the unattached marks in the Keyman On Screen Keyboard description.

The list of scripts to build is TBA, but anticipate that we would address all supported scripts over time.

### Details

The Keyman team have already created fonts to mitigate these issues for Khmer and Lao scripts. Thus, this solution has been tested and found to work reliably and straightforwardly. We'd like to apply this solution to other scripts for which we have seen these issues – including all scripts in South Asia and South-East Asia, Hebrew, Arabic. The full list of scripts is TBA.

Keyman has the ability to specify and embed an On Screen Keyboard font with each keyboard; this font is independent of the fonts used in text input.

For the Khmer script, we created a font [Khmer Busra Kbd](https://github.com/keymanapp/keyboards/blob/59c44e8384b736f5f8e51a0af39f91bc677f0cce/release/shared/fonts/khmer/busrakbd/khmer_busra_kbd.ttf) (khmer_busra_kbd.ttf) which extends an existing open source font, Khmer Busra. The Khmer Busra Kbd font includes a set of additional glyphs in the Private Use Area. These glyphs combine dotted circle with all the unattached marks and groups of marks that we need to display on the On Screen Keyboard, and so can be used without needing to rely on shaping from the renderer.

image

The work involved, for each script:

* Keyman team: Identify each of the marks required and assign a PUA codepoint to each one.
* Keyman team: Identify where a groups of marks may be required, e.g. 17C4 17C7 ោ ះ and assign a PUA codepoint to each group
* Keyman team: document assignments for use in the On Screen Keyboard (standard template to be built)
* Keyman team / WSTech team: Identify a suitable open source font to use as a basis
* Build the PUA codepoints from this font

Once the font is built, we can update the Keyman On Screen Keyboard descriptions for affected keyboards:
* Identify keyboards that have these issues 
* Update the On Screen Keyboard to use the PUA characters instead of the unattached bases

The new fonts do not need to have any OpenType or other shaping rules applied to the new PUA characters. Directionality is unimportant, because in all cases, the intention will be to display a single glyph centred on the key.

We would publish the binary fonts along with the documented assignments in the Keyman Keyboard Repository, for use by any keyboard author. It would be sensible to publish the source also in a GitHub repository for use in scenarios other than Keyman.

Fonts will be named according to a standard template, `Kbd` prefix, followed by a 4 letter script code, with an optional `-variant` such as `KbdArab-Nastaliq`. This would be the font face name as well as the font filename.

### Benefits

* On Screen Keyboard will always display consistently, regardless of platform
* Will work with LDML keyboards (as long as we can specify font association)
* These fonts could also be used for:
  * Technical documentation in Unicode and other specifications

### Implementation

`kmc analyze osk-char-use` tool will be used to generate a list of output USVs and for each one has a sequence of characters (USVs) that are shaped to produce the glyphs and positions. These are then composed to create the glyph that is mapped to the output USV. The generated font could then remove all unused glyphs and behaviors.

```
psfttf2osk -i input.ttf -o output.ttf -c charlist.json -n “KBDXxxx”
```

## References

* Discussion online - https://community.software.sil.org/t/hebrew-sil-keyboard-has-some-invisible-characters-on-ios-16-3-1-obu/7273/9
* Lao script research and solution - https://marc.durdin.net/2015/01/how-to-rendering-combining-marks-consistently-across-platforms-a-long-story/ 
* Khmer script solution - https://github.com/keymanapp/keyboards/pull/1710
* https://github.com/keymanapp/keyman/issues/1070 
* https://github.com/keymanapp/keyboards/issues/2172 
* Original discussion document (now archived):  https://docs.google.com/document/d/1gJrWpZebOlPn_yn91BGk_Z8hEvVDVdxXB0ihgkW7wdg/edit#heading=h.9kamffv3tas9

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #9031. 
Fixes #1070.
Builds on #9016.

The `&displayMap` system store adds support for a mapping file that remaps the On Screen Keyboard files -- .kvks and .keyman-touch-layout, at compile time, to new ranges. This is specifically intended to resolve a limitation of unattached marks and diacritics which display inconsistently across platforms; this is described in more detail in issue #9031.

In draft because this still needs work for KMW compiler -- but need to do some refactoring on KMW compiler, inverting dependency on kmc-kmn so that kmc-kmn calls into kmc-kmw, before this will be achievable.

Related PRs:
* https://github.com/keymanapp/keyboards/pull/2258
* https://github.com/keymanapp/help.keyman.com/pull/731
* #9016

TODO:
- [x] keyboards PR - https://github.com/keymanapp/keyboards/pull/2258
- [x] documentation - keyboard side - https://github.com/keymanapp/help.keyman.com/pull/731
- [x] documentation - process for creating custom osk font - https://github.com/keymanapp/help.keyman.com/pull/731
- [x] documentation - overview of problem and walkthrough of solution - #9031
- [x] documentation - json file format - in this PR
- [x] fill in #9031 from Google doc
- [x] linkage to python tooling for generating osk font - https://github.com/keymanapp/help.keyman.com/pull/731, also in this PR
- [x] kmc-kmw refactor to invert dep on kmc-kmn - #9038
- [x] kmc-kmw integration of this solution - #9038

@keymanapp-test-bot skip (testing will be in keyboards PR, once 17.0 tooling is in place)

### CODE REVIEW

File: developer/src/kmc-kmn/src/compiler/compiler.ts [hunk 1]
```diff
@@ -224,6 +226,26 @@ export class KmnCompiler implements UnicodeSetParser {
...
     }
+
+    if(displayMapFilename) {
+      // Remap using the osk-char-use-rewriter
+      let mapping: any;
+
+      displayMapFilename = this.callbacks.resolveFilename(kmnFilename, displayMapFilename);
+      try {
+        // Expected file format: displaymap.schema.json
+        // TODO: verify with schema
```
Comment (jahorton) [inline developer/src/kmc-kmn/src/compiler/compiler.ts:237]: I don't see a match for this among the PR's tracked items.  Is this being tracked somewhere/somehow?

File: developer/src/kmc-kmw/src/compiler/write-compiled-keyboard.ts [hunk 2]
```diff
@@ -210,6 +210,25 @@ export function WriteCompiledKeyboard(callbacks: CompilerCallbacks, kmnfile: str
       kvk = reader.read(callbacks.loadFile(path));
     }
 
+    // TODO: support &displayMap:
```
Comment (jahorton) [inline developer/src/kmc-kmw/src/compiler/write-compiled-keyboard.ts:213]: Tracked item:
> kmc-kmw integration of this solution

File: developer/src/kmc-kmw/test/test-compiler-manual.ts [hunk 3]
```diff
@@ -26,18 +26,19 @@ if(!await kmnCompiler.init(callbacks)) {
   process.exit(1);
 }
 
-// TODO: runToMemory, add option to kmxCompiler to store debug-data for conversion to .js (e.g. store metadata, group readonly metadata, etc)
-if(!kmnCompiler.run(infile, outfile, {
+let result = kmnCompiler.runCompiler(infile, outfile, {
   shouldAddCompilerVersion: false,
-  saveDebug: true,
+  saveDebug: true,  // TODO: we should probably use passed debug flag
```
Comment (jahorton) [inline developer/src/kmc-kmw/test/test-compiler-manual.ts:31]: Just pointing this one out.  I'm not sure if it's related to the following or not:

> kmc-kmw refactor to invert dep on kmc-kmn

**Review by jahorton — APPROVED** [review]
Saw a few things I just wanted to be sure are tracked in some fashion.  Aside from that, LGTM.

File: developer/src/kmc-kmw/test/test-compiler-manual.ts (hunk 3 above)
Comment (mcdurdin) [reply developer/src/kmc-kmw/test/test-compiler-manual.ts:31]: Yep, will be sorted out when I add proper tests for kmw compiler.

File: developer/src/kmc-kmn/src/compiler/compiler.ts (hunk 1 above)
Comment (mcdurdin) [reply developer/src/kmc-kmn/src/compiler/compiler.ts:237]: Good catch. Will add it here now so I don't forget -- it's a pretty simple addition.

**Pull request merged**